include/adc.h

```diff
@@ -27,119 +27,225 @@ extern "C" {
- * used to execute a sequence of samplings.
+ * @brief ADC sampling definition structure.
+ */
+struct adc_sequence {
+	/**
+	 * ADC resolution.
+	 * For single-ended channels the sample values are from range:
+	 *   0 .. 2^resolution - 1,
+	 * for differential ones:
+	 *  -2^(resolution-1) .. 2^(resolution-1) - 1.
+	 */
+	u8_t resolution;
+
+	/**
+	 * Oversampling setting.
+	 * Each sample is averaged from 2^oversampling results.
+	 */
+	u8_t oversampling;
+
+	/** Points to array of samplings to be performed in this sequence. */
+	const struct adc_sampling *samplings;
+
+	/**
+	 * Pointer to structure defining the way in which the sequence is to be
+	 * continued. If NULL, the sequence consists of a single sampling.
+	 */
+	const struct adc_sequence_continuation *continuation;
+};
+
+enum adc_prescaling
+{
+	ADC_PRESCALING_1_6,
+	ADC_PRESCALING_1_5,
+	ADC_PRESCALING_1_4,
+	ADC_PRESCALING_1_3,
+	ADC_PRESCALING_1_2,
+	ADC_PRESCALING_1,
+	ADC_PRESCALING_2,
+	ADC_PRESCALING_3,
+	ADC_PRESCALING_4,
+};
+
+enum adc_reference
+{
+	ADC_REF_VDD_1,
+	ADC_REF_VDD_1_2,
+	ADC_REF_VDD_1_3,
+	ADC_REF_VDD_1_4,
+	ADC_REF_INTERNAL,
+	ADC_REF_EXTERNAL0,
+	ADC_REF_EXTERNAL1,
+};
+
+#define ADC_ACQ_TIME_TICKS         (0u)
+#define ADC_ACQ_TIME_MICROSECONDS  (1u)
+#define ADC_ACQ_TIME_NANOSECONDS   (2u)
+#define ADC_ACQ_TIME(unit, value)  (((unit) << 14) | \
+				    ((value) & ((1u << 14)-1)))
+#define ADC_ACQ_TIME_UNIT(time)    (((time) >> 14) & 3u)
+#define ADC_ACQ_TIME_VALUE(time)   ((time) & ((1u << 14)-1)
+
+/**
+ * @brief ADC sampling definition structure.
  */
-struct adc_seq_table {
-	/* Pointer to a sequence entry array. */
-	struct adc_seq_entry *entries;
+struct adc_channel_cfg {
+	/** Channel identifier. */
+	u8_t channel_id;
+
+	/** Channel type: single-ended or differential. */
+	bool differential;
+
+	/** Positive converter input. */
+	u8_t input_positive;
+	/** Negative converter input (used only for differential channels). */
+	u8_t input_negative;
 
-	/* Number of entries in the sequence entry array. */
-	u8_t num_entries;
-	u8_t stride[3];
+	/**
+	 * Acquisition time.
+	 * Use the ADC_ACQ_TIME macro to compose the value for this field.
+	 * Particular drivers do not necessarily support all the possible units.
+	 * Value range is 0..16383 for a given unit.
+	 */
+	u16_t acquisition_time;
+
+	/** ADC reference selection. */
+	enum adc_reference  reference;
+	/** Channel voltage prescaling selection. */
+	enum adc_prescaling prescaling;
 };
 
 /**
  * @brief ADC driver API
  *
- * This structure holds all API function pointers.
+ * This is the mandatory API any ADC driver needs to expose.
  */
 struct adc_driver_api {
-	/** Pointer to the enable routine. */
-	void (*enable)(struct device *dev);
-
-	/** Pointer to the disable routine. */
-	void (*disable)(struct device *dev);
-
-	/** Pointer to the read routine. */
-	int (*read)(struct device *dev, struct adc_seq_table *seq_table);
+	int (*read)(struct device *dev,
```

Comment:
to be consistent  with other APIs, create function typedefs for the functions first